src/github.com/matrix-org/dendrite/clientapi/auth/storage/devices/storage.go

```diff
@@ -44,18 +44,15 @@ func NewDatabase(dataSourceName string, serverName gomatrixserverlib.ServerName)
 // GetDeviceByAccessToken returns the device matching the given access token.
 // Returns sql.ErrNoRows if no matching device was found.
 func (d *Database) GetDeviceByAccessToken(token string) (*authtypes.Device, error) {
-	// return d.devices.selectDeviceByToken(token) TODO: Figure out how to make integ tests pass
-	return &authtypes.Device{
-		UserID:      token,
-		AccessToken: token,
-	}, nil
+	return d.devices.selectDeviceByToken(token)
 }
 
 // CreateDevice makes a new device associated with the given user ID localpart.
 // If there is already a device with the same device ID for this user, that access token will be revoked
-// and replaced with a newly generated token.
+// and replaced with a newly generated token. If 'desiredAccessToken' is specified, the device will be
+// created with that access_token and an error will be returned if this is not possible.
```

Comment:
Hmm, maybe move the token generation up a layer? We could make it the responsibility of /login to generate the random token when logging in.